### ISSUE TITLE
[Suggestion] Option to enhance RSA

### ISSUE THREAD

**Issue by sant0ro** [ArachisH/HabKit#4]
The actual implementation way of Habbo about RSA is wrong.

We have the following in actual scenarion:
1. A communication tunnel under TCP with plain sockets, and usage of RC4 as single application layer protocol, used both for security and communication.
2. The server sends the DH Prime Numbers to the Client encrypted with RSA using the Public Shared Key that it's both stored in the Server and Client.
3. Client validates the Shared Public Key, decrypt the DH prime numbers that will be used as along the Habbo Encoding.
4. Communication between the Client <> Server is secure, using DH as salt for the Habbo Encoding.

The item 2, it's wrong on my opinion. The Client should have the Public Key and the Server the Private Key, but both uses the Private Keys. Client should validate the Public key with the Private Key, and after success continue the other steps. 

Actually a session between the User and Server it's secured if both RC4 and RSA and DH are enabled. But this doesn't prevents that PacketLoggers can connect to the Server.

For both TCP spooling and PacketLogger, Client should use the proper way of verification that the [RFC 3447](https://tools.ietf.org/html/rfc3447) specifies.

Since only the Emulator knows the Private Key, and PacketLoggers doesn't have access to him, PacketLoggers would be disconnected because they wouldn't have a valid Identification.

Actually PacketLoggers does the basics of communication interception, the famous MITM, basically a pakcet logger receives and injects packets along the communication betwwen Client <> Server.

Spoofing and Spooling would be more difficult, but PacketLogging not. It's a lot difficult handling over PacketLogging. For the Server Security, filtering, validation and checks **need** to be made. A good secure emulator it's Arcturus. But Client Side PacketLogging isn't impossible, but it's useless since it's Client Side.

What you think about this, @ArachisH ?

**Comment by jomp16** [comment]
"2. The server sends the DH Prime Numbers to the Client encrypted with 
RSA using the Public Shared Key that it's both stored in the Server and 
Client."

You're wrong, the server encrypts using the private key RSA and send 
back to the client, decrypting with the public key. And then the client 
generates a RC4 secret key and sends back to the server using the RSA 
public key, and the server decrypts it using it's RSA private key and 
generates a secret key, the final result is matching RC4 keys.

I've never saw a Habbo.swf using a private key, just a public key, that 
should be shared, because, you know, it's public?

I don't see anything needing to be fixed, it's working as intended.

About packetlogger, unfortunately I think we can't prevent them, since 
for example, Arachis's packet logger works by retrieving the Habbo.swf 
and injecting other public key and sending back to the browser. Using a 
HTTPS cert that the user installed to the OS, enabling MITM.

Unless I understood all of this wrong...

**Comment by sant0ro** [comment]
I think you misunderstood my explanation. Actually in Habbo both Client and Server are using Public Keys.

Actually in an Emulator RSA-enabled, stores the (m^e)^d, and the Client stores also the (m^e)^d, actually in the emulators that we use, we just see both doing checksum of both RSA keys.

I remember that any emulator finished the proper RSA way of implementation. Maybe I'm wrong, since its like 3 years that I don't touch the Crypto section of Emulators, and I don't exactly remember how it was implemented.

But if actually it's in the right way, so I'm sorry. But I remember that they weren't concluded, but again, I may be wrong.

**Comment by jomp16** [comment]
Just to make sure we're talking the same thing:

Private key: uses N and D

Public key: uses N and E

Client -> INIT CRYPTO

Server -> INIT CRYPYO - DH keys signed with RSA using private key

Client -> verifies keys using public key

Client -> SECRET KEY - signed with RSA using public key.

Server -> verifies keys using private key

(at this point, both shared keys is generated).

Is it right and excepted?

I should know this, because I worked on it on my server, using Java 
cipher (I needed to generate really big RSA keys to allow a really big 
DH param 512 bytes!). And I lost a lot of my time because I was 
verifying and signing with public, not private.

https://git.jomp16.tk/Habbo/habbo_r63b_v2/blob/master/server/src/main/kotlin/tk/jomp16/habbo/communication/incoming/handshake/HandshakeInitCryptoHandler.kt

https://git.jomp16.tk/Habbo/habbo_r63b_v2/blob/master/server/src/main/kotlin/tk/jomp16/habbo/encryption/HabboEncryptionHandler.kt

https://git.jomp16.tk/Habbo/habbo_r63b_v2/blob/master/server/src/main/kotlin/tk/jomp16/habbo/encryption/RSAEncryption.kt

**Comment by sant0ro** [comment]
Yes you're right, I confused things. Like I said, was years ago that I worked with this.

"I should know this, because I worked on it on my server, using Java
cipher (I needed to generate really big RSA keys to allow a really big
DH param 512 bytes!). And I lost a lot of my time because I was
verifying and signing with public, not private."

I think saying about this is unnecessary. I could also say that I'm a Network Communications Engineer..,. My questions wasn't about how RSA works, but if they was correctly implemented in actual emulators.

But yeah, now I remember that actually Joopie's HabboEncryption repo it's all right, and you're right, the communication works in this exactly way.

So, no problem here. My toughs were that our Emulators were doing it incorrectly.

Thanks for clarifying the questions 😄
